src/main/java/hudson/remoting/Engine.java

```diff
@@ -277,6 +290,10 @@ public void run() {
                         events.status("Protocol " + protocol.getName() + " is not enabled, skipping");
                         continue;
                     }
+                    if (agentProtocolNames != null && !agentProtocolNames.contains(protocol.getName())) {
```

Comment:
🐛 agentProtocolNames **will** be empty if X-Jenkins-Agent-Protocols is missing since you initialize the container before if clause. In such case this code will also skip protocols